lib/src/component_declaration/function_component.dart

```diff
@@ -0,0 +1,181 @@
+// Copyright 2020 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ignore_for_file: prefer_generic_function_type_aliases
+library over_react.component_declaration.function_component;
+
+import 'package:js/js_util.dart';
+import 'package:meta/meta.dart';
+import 'package:over_react/over_react.dart';
+import 'package:react/react.dart' as react;
+import 'package:react/react_client.dart';
+import 'package:react/react_client/js_backed_map.dart';
+
+export 'component_type_checking.dart'
+    show isComponentOfType, isValidElementOfType;
+
+/// A React component declared using a function that takes in [props] and returns rendered output.
+///
+/// See <https://facebook.github.io/react/docs/components-and-props.html#functional-and-class-components>.
+typedef dynamic UiFunctionComponent<T extends UiProps>(T props);
+
+/// Returns a `UiFactory<TProps>` for [functionComponent].
+///
+/// __Example__:
+/// ```dart
+/// UiFactory<FooProps> Foo = uiFunction((props) {
+///     // Set default props using null-aware operators.
+///     final isDisabled = props.isDisabled ?? false;
+///     final items = props.items ?? [];
+///
+///     // Return the rendered component contents here.
+///     return Dom.div()(items);
+///   },
+///   // The generated props config will match the factory name.
+///   $FooPropsConfig,
+/// );
+///
+/// mixin FooProps on UiProps {
+///   bool isDisabled;
+///   Iterable<String> items;
+/// }
+/// ```
+///
+/// __OR__ Optionally pass in an existing [PropsFactory] in place of a props [config].
+///
+/// ```dart
+/// UiFactory<FooProps> Bar = uiFunction((props) {
+///     return Dom.div()(props.items);
+///   },
+///   null,
+///   propsFactory: PropsFactory.fromUiFactory(Foo),
```

Comment:
I think we should also have an example of using an existing generated PropsConfig and a different display name, to show what declaring multiple function components with the same props looks like.

Reply:
I can add an example of multiple function components with the same props. I made it so a PropsConfig is generated for each function component so that the display name would be automatically set correctly:
```
mixin FooProps on UiProps {}

UiFactory<FooProps> Foo = uiFunction(
  (props) => Dom.div()(),
  $FooPropsConfig,
);

UiFactory<FooProps> Bar = uiFunction(
  (props) => Dom.div()(),
  $BarPropsConfig,
);
```
Does that make sense, or do you think I should change it so that one config is generated for each props mixin and then the display name has to be set on multiple function components? Or I could change the config names to `$FooConfig` and `$BarConfig` to associate them more with the factory than the props?

Reply:
Ooh nice, no, I like what you have a lot!

Reply:
Oh wait, I misread that last part.

> Or I could change the config names to `$FooConfig` and `$BarConfig` to associate them more with the factory than the props?

I think that makes sense, so that the naming aligns more with `FunctionComponentConfig` (it's not `FunctionComponentPropsConfig`) and doesn't get falsely associated with a props class of a given name (e.g., `BarProps` in `$BarPropsConfig`).

Reply:
Cool I'll update that!